aten/src/ATen/native/cuda/TensorTopK.cu

```diff
@@ -162,10 +171,342 @@ __global__ void gatherTopK(at::cuda::detail::TensorInfo<T, IndexType> input,
+      s_is_last_block_done = (blocks_finished_old == blocks_per_slice - 1);
+    }
+  }
+
+  __syncthreads();
+
+  if (!s_is_last_block_done)
+    return;
+
+  // accumulates counters from multiple blocks
+  if (tidx < RADIX_DIGITS && blocks_per_slice > 1) {
+    digit_count = 0;
+    for (int blk = 0; blk < blocks_per_slice; ++blk) {
+      digit_count += counts[(slice_idx * blocks_per_slice + blk) * RADIX_DIGITS + tidx];
+    }
+  }
+
+  // compute the block-wide inclusive prefix sum
+  IndexType& digit_count_cumsum = digit_count;
+  BlockPrefixCallbackOp prefix_op(0);
+  BlockScan(temp_storage.scan_storage).InclusiveSum(digit_count, digit_count_cumsum, prefix_op);
+  __syncthreads();
+  // every thread also need the perfix_sum of it's left value for comparison, so save a copy in shared mem
+  if (tidx < RADIX_DIGITS) {
+    temp_storage.digit_count_cumsum[tidx] = digit_count_cumsum;
+  }
+  __syncthreads();
+
+  if (tidx < RADIX_DIGITS) {
+    IndexType digit_count_cumsum_left = (tidx == 0) ? 0 : temp_storage.digit_count_cumsum[tidx - 1];
+
+    // if not the last pass: update desired and ks_to_find
+    // if last pass: write out the kth value
+    if (digit_count_cumsum_left < k_to_find && k_to_find <= digit_count_cumsum) {
+      desired = at::cuda::Bitfield<Bitwise>::setBitfield(desired, tidx, current_bit, RADIX_BITS);
+      if (current_bit > 0) {
+        desires[slice_idx] = desired;
+        ks_to_find[slice_idx] = k_to_find - digit_count_cumsum_left;
+      } else {
+        kthValues[slice_idx] = TopKTypeConfig<T>::deconvert(desired);
+      }
+    }
+  }
+
+  // reset semaphores for the next pass
+  if (tidx == 0) {
+    semaphores[slice_idx] = 0;
+  }
+};
+
+template <typename T, typename IndexType, int Dim>
+void launch(
+    at::cuda::detail::TensorInfo<T, IndexType> input,
+    IndexType inputSliceSize,
+    IndexType outputSliceSize, // aka `k`
+    bool largest,
+
+    IndexType numInputSlices,
+    IndexType inputWithinSliceStride,
+
+    at::cuda::detail::TensorInfo<T, IndexType> topK,
+    IndexType topKWithinSliceStride,
+
+    at::cuda::detail::TensorInfo<int64_t, IndexType> indices,
+    IndexType indicesWithinSliceStride) {
+
+  // configure items_per_thread based on device architecture and input size
+  // occupancy of this kernel is limited by registers per threads
+  constexpr int REGS_PER_THREAD = 40;
+  constexpr int REGS_PER_BLOCK = REGS_PER_THREAD * BLOCK_THREADS;
+  cudaDeviceProp* prop = at::cuda::getCurrentDeviceProperties();
+  int mpc = prop->multiProcessorCount;
+#if defined(USE_ROCM)
+  int regs_per_mp = prop->regsPerBlock;
+  int blocks_per_mp = regs_per_mp / REGS_PER_BLOCK;
+#else
+  int regs_per_mp = prop->regsPerMultiprocessor;
+  int blocks_per_mp = std::min(regs_per_mp / REGS_PER_BLOCK, prop->maxBlocksPerMultiProcessor);
+#endif
+  int items_per_thread = at::ceil_div((int64_t)(inputSliceSize * numInputSlices), (int64_t)(mpc * blocks_per_mp * BLOCK_THREADS));
```

Comment:
Should we use `int64_t` for this? In extreme case where `inputSliceSize * numInputSlices` is 2^64, `int` will not be sufficient to hold your value here.

Reply:
it make sense, fixed!